acceptance/github_asset_fetcher.go

```diff
@@ -0,0 +1,462 @@
+		if constraint.Check(versions[i]) {
+			latestPatchOfPreviousMinor = versions[i]
+			break
+		}
+	}
+	if latestPatchOfPreviousMinor == nil {
+		return "", errors.Wrapf(err, "obtaining latest patch of previous minor")
+	}
+	formattedVersion := fmt.Sprintf("v%s", latestPatchOfPreviousMinor.String())
+
+	err = f.writeCacheManifest(owner, repo, func(cache assetCache) {
+		cache[owner][repo].Versions[strconv.Itoa(n)] = formattedVersion
+	})
+	if err != nil {
+		f.logger.Warn(errors.Wrap(err, "writing cache").Error())
+	}
+	return formattedVersion, nil
+}
+
+func (f *GithubAssetFetcher) cachedAsset(owner, repo, version string, expr *regexp.Regexp) (string, error) {
+	cache, err := f.loadCacheManifest()
+	if err != nil {
+		return "", errors.Wrap(err, "loading cache")
+	}
+
+	assets, found := cache[owner][repo].Assets[version]
+	if found {
+		for _, asset := range assets {
+			if expr.MatchString(asset) {
+				return asset, nil
+			}
+		}
+	}
+	return "", nil
+}
+
+func (f *GithubAssetFetcher) cachedSource(owner, repo, version string) (string, error) {
+	cache, err := f.loadCacheManifest()
+	if err != nil {
+		return "", errors.Wrap(err, "loading cache")
+	}
+
+	value, found := cache[owner][repo].Sources[version]
+	if found {
+		return value, nil
+	}
+	return "", nil
+}
+
+func (f *GithubAssetFetcher) cachedVersion(owner, repo string, n int) (string, error) {
+	cache, err := f.loadCacheManifest()
+	if err != nil {
+		return "", errors.Wrap(err, "loading cache")
+	}
+
+	value, found := cache[owner][repo].Versions[strconv.Itoa(n)]
+	if found {
+		return value, nil
+	}
+	return "", nil
+}
+
+func (f *GithubAssetFetcher) loadCacheManifest() (assetCache, error) {
+	cacheManifest, err := os.Stat(filepath.Join(f.cacheDir, assetCacheManifest))
+	if os.IsNotExist(err) {
+		return assetCache{}, nil
+	}
+
+	// invalidate cache manifest that is too old
+	if time.Since(cacheManifest.ModTime()) > 1*time.Hour {
+		return assetCache{}, nil
+	}
+
+	content, err := ioutil.ReadFile(filepath.Join(f.cacheDir, assetCacheManifest))
+	if err != nil {
+		return nil, errors.Wrap(err, "reading cache manifest")
+	}
+
+	var cache assetCache
+	err = json.Unmarshal(content, &cache)
+	if err != nil {
+		return nil, errors.Wrap(err, "unmarshaling cache manifest content")
+	}
+
+	return cache, nil
+}
+
+func (f *GithubAssetFetcher) writeCacheManifest(owner, repo string, op func(cache assetCache)) error {
+	cache, err := f.loadCacheManifest()
+	if err != nil {
+		return errors.Wrap(err, "loading cache")
+	}
+
+	// init keys for owner and repo
+	if _, found := cache[owner]; !found {
+		cache[owner] = map[string]cachedRepo{}
+	}
+	if _, found := cache[owner][repo]; !found {
+		cache[owner][repo] = cachedRepo{
+			Assets:   cachedAssets{},
+			Sources:  cachedSources{},
+			Versions: cachedVersions{},
+		}
+	}
+
+	op(cache)
+
+	content, err := json.Marshal(cache)
+	if err != nil {
+		return errors.Wrap(err, "marshaling cache manifest content")
+	}
+	content = append(content, "\n"...)
```

Comment:
Why are we appending a new line?

Reply:
I think this is an artifact of my debugging - will take this out.